gpm_api/tests/io/test_filter.py

```diff
@@ -0,0 +1,266 @@
+        start_time=datetime.datetime(2019, 1, 1),
+        end_time=datetime.datetime(2019, 12, 31, 23, 59, 59),
+        version=7,
+    )
+    assert res == []
+
+    # Test empty filepath list
+    res = filter.filter_filepaths(
+        filepaths=[],
+        product="2A-DPR",
+        start_time=datetime.datetime(2019, 1, 1),
+        end_time=datetime.datetime(2019, 12, 31, 23, 59, 59),
+        version=7,
+    )
+    assert res == []
+
+    # Test empty start time
+    count_until_2019 = 0
+    for server_path, props in server_paths.items():
+        if props["year"] == 2019:
+            count_until_2019 += 1
+    res = filter.filter_filepaths(
+        filepaths=list(server_paths.keys()),
+        product="2A-DPR",
+        start_time=None,
+        end_time=datetime.datetime(2019, 12, 31, 23, 59, 59),
+        version=7,
+    )
+
+    assert len(res) == count_until_2019
+
+    # Test empty end time (Error as time given (datetime.datetime.now())
+    # requires date to be less than now() in supportive
+    # function checks.check_start_end_time)
+    count_from_2019 = 0
+    for server_path, props in server_paths.items():
+        if props["year"] >= 2019:
+            count_from_2019 += 1
+
+    res = filter.filter_filepaths(
+        filepaths=list(server_paths.keys()),
+        product="2A-DPR",
+        start_time=datetime.datetime(2019, 1, 1),
+        end_time=None,
+        version=7,
+    )
+    assert len(res) == count_from_2019
+
+
+def test_filter_by_time(
+    server_paths: Dict[str, Dict[str, Any]],
+) -> None:
+    """Test filter filepaths"""
+
+    # Test year filtering
+    # Count and assert 2019 paths
+    count_2019 = 0
+    for server_path, props in server_paths.items():
+        if props["year"] == 2019:
+            count_2019 += 1
+
+    res = filter.filter_by_time(
+        filepaths=list(server_paths.keys()),
+        start_time=datetime.datetime(2019, 1, 1),
+        end_time=datetime.datetime(2019, 12, 31, 23, 59, 59),
+    )
+
+    assert len(res) == count_2019
+
+    # Test None filepaths
+    res = filter.filter_by_time(
+        filepaths=None,
+        start_time=datetime.datetime(2019, 1, 1),
+        end_time=datetime.datetime(2019, 12, 31, 23, 59, 59),
+    )
+
+    assert res == []
+
+    # Test empty filepath list
+    res = filter.filter_by_time(
+        filepaths=[],
+        start_time=datetime.datetime(2019, 1, 1),
+        end_time=datetime.datetime(2019, 12, 31, 23, 59, 59),
+    )
+
+    assert res == []
+
+    # Test empty start time
+    count_until_2019 = 0
+    for server_path, props in server_paths.items():
+        if props["year"] == 2019:
+            count_until_2019 += 1
+    res = filter.filter_by_time(
+        filepaths=list(server_paths.keys()),
+        start_time=None,
+        end_time=datetime.datetime(2019, 12, 31, 23, 59, 59),
+    )
+
+    assert len(res) == count_until_2019
+
+    # TODO: Investigate why this fails in GH Actions but passes locally (system time issue?)
```

Comment:
Could you link the corresponding Github action for us to have a look?

Reply:
https://github.com/ghiggi/gpm_api/actions/runs/5873832307/job/15927650742#step:5:258

As I understand, if there is a `None` end time, the function will set the time to now(), but this will mean that `check_start_end_time` throws an error that the time needs to be in the past. Have you run this locally?

[def check_start_end_time(start_time, end_time):](https://github.com/ghiggi/gpm_api/blob/9f2cd9bdd70b7b83d3b88f5d3847cdcd8b6589e0/gpm_api/io/checks.py#L188)

Reply:
It works on my laptop. It could be a timezone issue? In `io/filter.py:263` we have `datetime.datetime.now()`, whereas in `io/checks.py:64` it's `datetime.datetime.utcnow()`

Reply:
You've a good eye, I think you may be correct, it looks like GH actions uses UTC time, so there would be no difference between `utcnow()` and `now()`. If I change my local timezone to UTC it will also fail.

I am going to write a couple of new tests for check_start_end_time. I think it would be ideal to have a consistent use of what the current time is and it may be worth changing the `now()` to `utcnow()`. What do you think?

```
>>> print(datetime.datetime.now().isoformat(), '|', datetime.datetime.utcnow().isoformat())
2023-08-21T09:16:07.168004 | 2023-08-21T07:16:07.168031
```

Reply:
As a user, I think that it would be better to have the time default to their current timezone if not specified, that would be `.now()`,

Reply:
The time on all the GPM products are [defined in UTC](https://gpm.nasa.gov/sites/default/files/2020-02/FileNamingConventionForPrecipitationProductsForGPMMission.pdf). In this case, the user should be entering UTC times in their queries.

Reply:
Okay in that case UTC is indeed the way to go :)  Maybe add a description of the default parameter in the docstrings

Reply:
Hey :) Sorry for my delayed answers ... are very busy days. We should always working with UTC ! If you find occurrence of now() instead of utcnow() ... replace it ... it's a bug !

Reply:
I've made the change on `now()` to `utcnow()` and added a couple of tests on these assumptions of time in `gpm_api.tests.io.test_checks.test_check_start_end_time()`